cmd/mt-whisper-importer-reader/conversion.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"github.com/kisielk/whisper-go/whisper"
+	"github.com/raintank/metrictank/mdata"
+)
+
+type conversion struct {
+	archives []whisper.ArchiveInfo
+	points   map[int][]whisper.Point
+	method   string
+}
+
+func newConversion(arch []whisper.ArchiveInfo, points map[int][]whisper.Point, method string) *conversion {
+	return &conversion{archives: arch, points: points, method: method}
+}
+
+func (c *conversion) findSmallestLargestArchive(spp, nop uint32) (int, int) {
+	// find smallest archive that still contains enough data to satisfy requested range
+	largestArchiveIdx := len(c.archives) - 1
+	for i := largestArchiveIdx; i >= 0; i-- {
+		arch := c.archives[i]
+		if arch.Points*arch.SecondsPerPoint < nop*spp {
+			break
+		}
+		largestArchiveIdx = i
+	}
+
+	// find largest archive that still has higher resolution than requested
+	smallestArchiveIdx := 0
+	for i := 0; i < len(c.archives); i++ {
+		arch := c.archives[i]
+		if arch.SecondsPerPoint > spp {
+			break
+		}
+		smallestArchiveIdx = i
+	}
+
+	return smallestArchiveIdx, largestArchiveIdx
+}
+
+// generates points according to specified parameters by finding and using the best archives as input
+func (c *conversion) getPoints(retIdx int, spp, nop uint32) map[string][]whisper.Point {
+	res := make(map[string][]whisper.Point)
+
+	if len(c.points) == 0 {
+		return res
+	}
+
+	// figure out the range of archives that make sense to use for the requested specs
+	smallestArchiveIdx, largestArchiveIdx := c.findSmallestLargestArchive(spp, nop)
+	rawRes := c.archives[0].SecondsPerPoint
+
+	adjustedPoints := make(map[string]map[uint32]float64)
+	if retIdx > 0 && c.method == "avg" {
+		adjustedPoints["cnt"] = make(map[uint32]float64)
+		adjustedPoints["sum"] = make(map[uint32]float64)
+	} else {
+		adjustedPoints[c.method] = make(map[uint32]float64)
+	}
+
+	// Out of the input archives that we'll use, start with the lowest resolution one by converting
+	// it to the requested resolution and filling the resulting points into adjustedPoints.
+	// Then continue with archives of increasing resolutions while overwriting the generated points
+	// of previous ones.
+	for i := largestArchiveIdx; i >= smallestArchiveIdx; i-- {
+		in := c.points[i]
+		arch := c.archives[i]
+		if arch.SecondsPerPoint == spp {
+			if retIdx == 0 || c.method != "avg" {
+				for _, p := range in {
+					adjustedPoints[c.method][p.Timestamp] = p.Value
+				}
+			} else {
+				for _, p := range in {
+					adjustedPoints["sum"][p.Timestamp] = p.Value
+					adjustedPoints["cnt"][p.Timestamp] = 1
+				}
+			}
+		} else if arch.SecondsPerPoint > spp {
+			method := c.method
+			if method == "avg" && retIdx > 0 {
+				method = "fakeavg"
```

Comment:
just so that i understand this, the archives submitted to the writer will have "fakeavg" method, but the method will be pretty much ignored by the writer, right? it'll just write the chunks to cassandra based on their id/key, not using the method information?

Reply:
The writer doesn't know how the data it receives has been generated. It only gets data of the type `Metric` (https://github.com/raintank/metrictank/blob/master/mdata/chunk/archive/archive.go#L24) and the aggregation method defined in the property `AggregationMethod` is not relevant for the way how we store the data in MT.
 The variable `method` is only used to tell the function `incResolution()` how to increase the resolution, because depending on the aggregation method it needs to do that slightly differently.
The writer then basically just gets chunks and rowkeys and writes them into cassandra,